src/NHapi.Base/Model/GenericMessage.cs

```diff
@@ -1,210 +1,224 @@
-		/// Version 2.4 generic message
-		/// </summary>
-		public class V24 : GenericMessage
-		{
-			/// <summary>
-			/// Version of message
-			/// </summary>
-			public override String Version
-			{
-				get { return "2.4"; }
-			}
-
-			/// <summary>
-			/// Constructor
-			/// </summary>
-			/// <param name="factory"></param>
-			public V24(IModelClassFactory factory)
-				: base(factory)
-			{
-			}
-		}
-
-		/// <summary>
-		/// Version 2.5 generic message
-		/// </summary>
-		public class V25 : GenericMessage
-		{
-			/// <summary>
-			/// Version of message
-			/// </summary>
-			public override String Version
-			{
-				get { return "2.5"; }
-			}
-
-			/// <summary>
-			/// Constructor
-			/// </summary>
-			/// <param name="factory"></param>
-			public V25(IModelClassFactory factory)
-				: base(factory)
-			{
-			}
-		}
-	}
+    using System;
+
+    using NHapi.Base;
+    using NHapi.Base.Log;
+    using NHapi.Base.Parser;
+
+    /// <summary> A generic HL7 message, meant for parsing message with unrecognized structures
+    /// into a flat list of segments.
+    /// </summary>
+    /// <author>  Bryan Tripp.
+    /// </author>
+    public abstract class GenericMessage : AbstractMessage
+    {
+        /// <summary> Creates a new instance of GenericMessage.
+        ///
+        /// </summary>
+        /// <param name="factory">class factory for contained structures.
+        /// </param>
+        public GenericMessage(IModelClassFactory factory)
+            : base(factory)
+        {
+            try
+            {
+                AddNonstandardSegment("MSH");
+            }
+            catch (HL7Exception e)
+            {
+                var message = "Unexpected error adding GenericSegment to GenericMessage.";
+                HapiLogFactory.GetHapiLog(GetType()).Error(message, e);
+                throw new ApplicationException(message);
+            }
+        }
+
+        [Obsolete("This method has been replaced by 'GetGenericMessageClass'.")]
+        [System.Diagnostics.CodeAnalysis.SuppressMessage(
+            "StyleCop.CSharp.NamingRules",
+            "SA1300:Element should begin with upper-case letter",
+            Justification = "As this is a public member, we will duplicate the method and mark this one as obsolete.")]
+        public static Type getGenericMessageClass(string version)
+        {
+            return GetGenericMessageClass(version);
+        }
+
+        /// <summary> Returns a subclass of GenericMessage corresponding to a certain version.
+        /// This is needed so that version-specific segments can be added as the message
+        /// is parsed.
+        /// </summary>
+        public static Type GetGenericMessageClass(string version)
+        {
+            if (!ParserBase.ValidVersion(version))
+            {
+                throw new ArgumentException("The version " + version + " is not recognized");
+            }
+
+            Type c = null;
+            if (version.Equals("2.1"))

```

Comment:
you might be able to use the new fancy [switch expression](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/switch-expression) pattern match stuff here... if it doesnt break net3.5 support

Reply:
have left a `TODO` comment for now

Reply:
Okay, thats fine.